src/algebra/algebra/non_unital.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Non-unital, non-associative algebras
+
+This files defines non-unital, non-associative algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Main definitions
+
+  * `non_unital_non_assoc_algebra`
+  * `non_unital_non_assoc_algebra_hom`
+
+## Tags
+
+non-unital, non-associative, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
+
+/-- A not-necessarily-unital, not-necessarily-associative algebra. -/
+@[protect_proj]
+class non_unital_non_assoc_algebra [non_unital_non_assoc_semiring A] extends semimodule R A :=
+(smul_mul_assoc' : ∀ (t : R) (a b : A), (t • a) * b = t • (a * b))
+(mul_smul_comm' : ∀ (t : R) (a b : A), a * (t • b) = t • (a * b))
+
+@[priority 200] -- see Note [lower instance priority]
+instance semiring.to_non_unital_non_assoc_semiring [semiring A] : non_unital_non_assoc_semiring A :=
+{ ..(infer_instance : semiring A) }
+
+@[priority 200] -- see Note [lower instance priority]
+instance algebra.to_non_unital_non_assoc_algebra (R : Type u₁)
+  [comm_semiring R] [semiring A] [algebra R A] : non_unital_non_assoc_algebra R A :=
+{ smul_mul_assoc' := algebra.smul_mul_assoc,
+  mul_smul_comm'  := algebra.mul_smul_comm,
+  ..algebra.to_semimodule }
+
+namespace non_unital_non_assoc_algebra
+
+variables {R A} [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+
+lemma smul_mul_assoc (t : R) (a b : A) :
+  (t • a) * b = t • (a * b) :=
+non_unital_non_assoc_algebra.smul_mul_assoc' t a b
+
+lemma mul_smul_comm (t : R) (a b : A) :
+  a * (t • b) = t • (a * b) :=
+non_unital_non_assoc_algebra.mul_smul_comm' t a b
+
+/-- The algebra multiplication as a bilinear map. -/
+def mul_as_bilinear {R : Type u₁} {A : Type u₂}
```

Comment:
We call this `algebra.lmul` in the normal case, right? Can we use the same convention here?